docs/DOCUMENTATION.md

```diff
@@ -0,0 +1,872 @@
+- There is a minimum number of submissions which has to be reached in order for the on-chain price to be updated
+- An oracle cannot initialise multiple consecutive rounds to avoid network congestion
+- The updated price is the median from the submitted prices for a particular round
+- Submissions for old rounds cannot be submitted and will be rejected
+
+## Aims/Objectives
+
+#### Middleware
+
+The middleware is needed to:
+1. Create CL jobs. Chainlink Jobs are created every minute. A cron job is not used in order to be able to pass in particular inputs to the job run so that the progress of the request can be stored and monitored.
+2. Get results back from the CL node.
+3. Query smart contract for new rounds.
+3. Push price updates on chain.
+
+The middleware should contain the following functionalities:
+
+1. A state file which is updated so that whenever the middleware is restarted it is able to continue from where it was stopped.
+2. An endpoint to listen for new and removed jobs from the CL node. This is needed in order to keep a list of jobs for which the middleware has to send requests.
+3. Keep a list of these jobs in its state
+4. Query the price on chain and the latest round every X seconds so that a new job request is created if a new round is found
+5. Send CL job requests to the CL node with parameters including a request ID and the reason for the request every Y seconds. The reason can be one of the following 3 reasons:
+  a. Type 1: Time expired and a new request have to be made
+  b. Type 2: A price deviation from the on-chain price was found
+  c. Type 3: A new round was found
+6. An endpoint to listen for results from a CL job for submitted job requests. This endpoint should decide whether or not to push the price on chain. A price should be pushed on chain if one of the following cases is satisfied:
+  a. It is a new round and the oracle has not yet submitted to this round
+  b. An interval of Z minutes/hours expired since the last price pushed on chain. This is only done if the oracle has not started the previous round himself. This is done because an oracle is not allowed to start consecutive rounds.
+  c. There is a deviation of more than W% from the current price on-chain. Once again, only done if the oracle has not started the previous round himself.
+7. Keep the following information in the state file for each feed:
+  a. The CL job's external id for that feed
+  b. The name of the feed
+  c. The last request id
+  d. The last reported round on chain
+  e. The timestamp of the last request made to the CL node
+  f. The last aggregated price on chain
+  g. The request id of the last result received from the CL node
+  h. The details of the latest on-chain round including the round id, the timestamp when it was started, by whom it was started and whether the oracle made a submission for that round
```

Comment:
Better doing this in a table, field name, struct and description

Reply:
Fixed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/d85360c17cb9f44394ff03ce5df97da7c44ef58a